core/matrix/ell.hpp

```diff
@@ -0,0 +1,274 @@
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CORE_MATRIX_ELL_HPP_
+#define GKO_CORE_MATRIX_ELL_HPP_
+
+
+#include "core/base/array.hpp"
+#include "core/base/convertible.hpp"
+#include "core/base/lin_op.hpp"
+#include "core/base/mtx_reader.hpp"
+
+
+namespace gko {
+namespace matrix {
+
+
+template <typename ValueType>
+class Dense;
+
+
+/**
+ * ELL is a matrix format where padding with explicit zeros is used such that
+ * all rows have the same number of stored elements. The number of elements
+ * stored in each row is the largest number of nonzero elements in any of the
+ * rows (obtainable through get_max_nonzeros_per_row() method). This removes
+ * the need of a row pointer like in the CSR format, and allows for SIMD
+ * processing of the distinct rows. For efficient processing, the nonzero
+ * elements and the corresponding column indices are stored in column-major
+ * fashion. The columns are padded to the length by user-defined padding
+ * parameter whose default value is the number of rows of the matrix.
+ *
+ * @tparam ValueType  precision of matrix elements
+ * @tparam IndexType  precision of matrix indexes
+ *
+ */
+template <typename ValueType = default_precision, typename IndexType = int32>
+class Ell : public BasicLinOp<Ell<ValueType, IndexType>>,
+            public ConvertibleTo<Dense<ValueType>>,
+            public ReadableFromMtx {
+    friend class BasicLinOp<Ell>;
+    friend class Dense<ValueType>;
+
+public:
+    using BasicLinOp<Ell>::create;
+    using BasicLinOp<Ell>::convert_to;
+    using BasicLinOp<Ell>::move_to;
+
+    using value_type = ValueType;
+    using index_type = IndexType;
+
+    void apply(const LinOp *b, LinOp *x) const override;
+
+    void apply(const LinOp *alpha, const LinOp *b, const LinOp *beta,
+               LinOp *x) const override;
+
+    void convert_to(Dense<ValueType> *other) const override;
+
+    void move_to(Dense<ValueType> *other) override;
+
+    void read_from_mtx(const std::string &filename) override;
+
+    /**
+     * Returns the values of the matrix.
+     *
+     * @return the values of the matrix.
+     */
+    value_type *get_values() noexcept { return this->values_.get_data(); }
+
+    /**
+     * @copydoc Ell::get_values()
+     *
+     * @note This is the constant version of the function, which can be
+     *       significantly more memory efficient than the non-constant version,
+     *       so always prefer this version.
+     */
+    const value_type *get_const_values() const noexcept
+    {
+        return this->values_.get_const_data();
```

Comment:
Ah, sorry, probably didn't make it clear. You need `this->` only when calling methods, not when accessing data members. So it's: `values_.get_const_data()`, but `this->get_cold_idxs()`. For data members the `_` at the end tells you it is a member of `this`.

Reply:
fixed

Reply:
@yanjen  also in other methods. Other than that, I'm really happy with this PR, and we can merge once this is done.

Reply:
I have checked again and removed the other parts

Reply:
@gflegar @hartwiganzt thank you for your kind help and patient.